### PR TITLE
Add Intervals class to allow checking for subsets and intersections between constraints

### PR DESCRIPTION
**Pull request by Seldaek**

### CODE REVIEW

File: src/Constraint/ConstraintInterface.php [hunk 1]
```diff
@@ -11,15 +11,31 @@
 
 namespace Composer\Semver\Constraint;
 
+/**
+ * DO NOT IMPLEMENT this interface. It is only meant for usage as a type hint
+ * where appropriate but we do not support third parties implementing this
+ * interface themselves, and will do BC breaks to the interface as we see fit.
+ */
```
Comment (staabm) [inline src/Constraint/ConstraintInterface.php:18]:
```suggestion
 * @internal
 */
```

Comment (Seldaek) [reply src/Constraint/ConstraintInterface.php:18]: Yeah my concern with `@internal` was that IDEs would complain if the interface is used as type hint outside the package. If that's not the case I happily add that though. I just have no idea how PHPStorm behaves exactly.

File: src/Constraint/Constraint.php [hunk 2]
```diff
@@ -84,6 +106,42 @@ public function matches(ConstraintInterface $provider)
...
+        if ($constraint instanceof MultiConstraint) {
+            if ($constraint->isConjunctive()) {
+                foreach ($constraint->getConstraints() as $c) {
+                    if (!$this->isSubsetOf($c)) {
+                        return false;
+                    }
+                }
+
+                return true;
+            } else {
```
Comment (stof) [inline src/Constraint/Constraint.php:127]: no need to use `else` as the `if` returns.

File: src/Constraint/Constraint.php [hunk 3]
```diff
@@ -84,6 +106,42 @@ public function matches(ConstraintInterface $provider)
...
+
+                return false;
+            }
+        }
+
+        if ($this->operator === self::OP_NE) {
+            return (string) $constraint === (string) $this;
+        }
+
+        return $constraint->matches($this);
```
Comment (stof) [inline src/Constraint/Constraint.php:142]: are you sure about that ? The fact that they intersect does not make it a subset.

File: src/Constraint/MultiConstraint.php [hunk 4]
```diff
@@ -132,6 +132,34 @@ public function matches(ConstraintInterface $provider)
...
+    /**
+     * {@inheritDoc}
+     */
+    public function isSubsetOf(ConstraintInterface $constraint)
+    {
+        if ($constraint instanceof EmptyConstraint) {
+            return true;
+        }
+
+        if (true === $this->conjunctive) {
```
Comment (stof) [inline src/Constraint/MultiConstraint.php:144]: no need for `true ===`. `$this->conjunctive` is already a boolean.

File: src/Constraint/MultiConstraint.php [hunk 5]
```diff
@@ -132,6 +132,34 @@ public function matches(ConstraintInterface $provider)
...
+            return true;
+        }
+
+        foreach ($this->getConstraints() as $c) {
+            if ($c->isSubsetOf($constraint)) {
+                return true;
+            }
+        }
+
+        return false;
```
Comment (stof) [inline src/Constraint/MultiConstraint.php:160]: this looks incomplete to me, in case the MultiConstraint is not optimized to merge ranges.

File: src/Constraint/Constraint.php (hunk 3 above)
Comment (stof) [reply src/Constraint/Constraint.php:142]: See `<2.0` and `>=1.1` for an example. These constraints are intersecting, but they are not a subset of each other.

Comment (Seldaek) [reply src/Constraint/Constraint.php:142]: Yeah no nothing is sure here, it was a first draft rough implementation but it is very much broken :)

File: src/Constraint/Constraint.php [hunk 6]
```diff
@@ -84,6 +106,42 @@ public function matches(ConstraintInterface $provider)
...
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+        }
+
+        if ($this->operator === self::OP_NE) {
+            return (string) $constraint === (string) $this;
```
Comment (jderusse) [inline src/Constraint/Constraint.php:139]:
```suggestion
            return $constraint->operator === self::OP_NE && $constraint->version === $this->version && $;
```

Comment (Seldaek) [reply src/Constraint/Constraint.php:139]: @jderusse I'm working on a bunch more changes locally so maybe wait before spending too much time reviewing/suggesting things here :)

Comment (stof) [reply src/Constraint/Constraint.php:139]: you suggestion looks like invalid code.

Comment (jderusse) [reply src/Constraint/Constraint.php]: thanks

File: src/Constraint/ConstraintInterface.php (hunk 1 above)
Comment (Toflar) [reply src/Constraint/ConstraintInterface.php:18]: I think the message is misleading. You should be type hinting towards this interface and thus `and will do BC breaks to the interface as we see fit` is incorrect. What you shouldn't do is invent your own constraint that implements this interface. I'd write something along the lines of
```
 * DO NOT IMPLEMENT this interface. It is only meant for usage as a type hint
 * in libraries relying on composer/semver but creating your own constraint class
 * that implements this interface is not a supported use case and will cause the
 * composer/semver components to return unexpected results.
```

Comment (Toflar) [reply src/Constraint/ConstraintInterface.php:18]: I know we're technically doing `BC breaks to the interface as we see fit` by adding a method like you did. But we cannot just remove one, that's why I wouldn't write that :)

Comment (Seldaek) [reply src/Constraint/ConstraintInterface.php:18]: @Toflar fair point thanks

File: src/Constraint/Constraint.php [hunk 7]
```diff
@@ -69,6 +69,38 @@ class Constraint implements CompilableConstraintInterface
     /** @var Bound */
     protected $upperBound;
 
+    /**
+     * Sets operator and version to compare with.
+     *
+     * @param string $operator
```
Comment (staabm) [inline src/Constraint/Constraint.php:75]:
```suggestion
     * @param string $operator
     * @psalm-param self::OP_*
```

Comment (staabm) [reply src/Constraint/Constraint.php:75]: More phpdocs in a separate PR https://github.com/composer/semver/pull/98

File: tests/IntervalsTest.php [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+            'conjunctive constraints should be intersected/2' => array(
+                array('intervals' => array(
+                    array(
+                        'start' => '>= 1.5.0.0-dev',
+                        'end' => '< 1.7.0.0-dev',
+                    ),
+                ), 'devConstraints' => array()),
+                '^1.0, ^1.2, 1.4 - 1.8, 1.5 - 1.6, 1.5 - 2'
+            ),
+            'conjunctive constraints should be intersected/not flattened by version parser' => array(
```
Comment (naderman) [inline tests/IntervalsTest.php:224]: What does "flattened" mean in this context?

Comment (Seldaek) [reply tests/IntervalsTest.php:224]: This tests exactly the same as the above, but the VersionParser takes all constraints and makes a single MultiConstraint out of them all as it's a conjunctive list.. So wanted to make sure the nested/"not flattened" multiconstraints also works.

Comment (naderman) [reply tests/IntervalsTest.php:224]: Oh I see, the way it was written I thought somehow "not flattened" was an alterntive way to say "intersected", guess you could replace the "/" with an "and", but all good ;-)

File: src/Intervals.php [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+{
+    private static $intervalsCache = array();
+
+    /**
+     * @psalm-var array<string, int>
+     */
+    private static $opSortOrder = array(
+        '>=' => -2,
+        '<' => -1,
+        '==' => 0,
```
Comment (naderman) [inline src/Intervals.php:32]: Can drop `==` here as they get converted

File: src/Intervals.php [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+            if ($border['side'] === 'start') {
+                $activeIntervals++;
+            } else {
+                $activeIntervals--;
+            }
+            if (!$active && $activeIntervals >= $activationThreshold) {
+                $intervals[] = array('start' => new Constraint($border['operator'], $border['version']));
+                $active = $activeIntervals;
+            }
+            if ($active && $activeIntervals < $active) {
```
Comment (naderman) [inline src/Intervals.php:167]: Shouldn't this be `<$activationThreshold`? No need to store the number in `$active`?

Comment (Seldaek) [reply src/Intervals.php]: Yes true, got sidetracked with another bug at some point, but this isn't needed :)

**Comment by Seldaek** [comment]
Ok all tests passing except one now, which is technically a valid result but not really what I'd like it to return.. @Toflar if you'd like to try and use this latest version in your Composer PR feel free.

File: src/Intervals.php [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+            if ((string) $c !== (string) $candidateIntervals['devConstraints'][$index]) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * @return array
```
Comment (staabm) [inline src/Intervals.php:82]:
```suggestion
     * @return array
     * @phpstan-return array{'intervals': array{'start': Constraint, 'end': Constraint}, 'devConstraints': Constraint[]}
```

**Comment by Seldaek** [comment]
All tests now passing.

https://github.com/composer/semver/pull/97/commits/5aef5c27d3f30e68651945bf7e2bfb817f9e46df adds a new isIntersectionOf method which could theoretically replace ConstraintInterface::matches() but I doubt it's as efficient so maybe not. We need to test and see. Anyway it was a great experiment as it allowed me to fix a few more edge cases in the intervals generation.

@naderman would be happy to get your opinion on https://github.com/composer/semver/pull/97/commits/67b8bb55c5426b0c3bf9e20e8e69b9972c92f4d6 and https://github.com/composer/semver/pull/97/commits/0b0edb398d272be11774c223dc3e2d030d7898d3 especially

File: src/Intervals.php [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+        foreach ($intersectionIntervals['devConstraints'] as $index => $c) {
+            if ((string) $c !== (string) $candidateIntervals['devConstraints'][$index]) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    public static function isIntersectionOf(ConstraintInterface $a, ConstraintInterface $b)
```
Comment (stof) [inline src/Intervals.php:81]: the name looks weird to me. This checks whether the 2 constraints are intersecting (they have a non empty intersection), not whether one of them is an intersection of the other (that does not make any sense).

Comment (naderman) [reply src/Intervals.php:81]: indeed, `intersectsWith` is probably better

Comment (Seldaek) [reply src/Intervals.php:81]: Yup true, I had it as `intersects` at first but then thought it sounded weird next to isSubsetOf. Then we also have `get` which I guess should be `getIntervals` for clarity as this grew to have more than just one getter.

I'd say `haveIntersection` if we return bool probably makes sense here. or `isIntersecting` as in `$a is intersecting $b`? The plural form kinda sounds odd.

Comment (naderman) [reply src/Intervals.php:81]: The class is plural `Intervals` so I think the most logical sounding is `lntervals::haveIntersections` ?

File: src/Intervals.php [hunk 13]
```diff
@@ -0,0 +1,369 @@
...
+            if (!$active && $activeIntervals >= $activationThreshold) {
+                $intervals[$index] = array('start' => new Constraint($border['operator'], $border['version']));
+                $active = true;
+            }
+            if ($active && $activeIntervals < $activationThreshold) {
+                $active = false;
+
+                // filter out invalid intervals like > x - <= x, or >= x - < x
+                if (
+                    version_compare($intervals[$index]['start']->getVersion(), $border['version'], '=')
```
Comment (naderman) [inline src/Intervals.php:277]: Is it necessary to use version_compare even with normalized versions for equality?

**Comment by naderman** [comment]
I still think it'd be better to have an actual Interval class rather than these arrays with start/end attributes which essentially act like an untyped object.

File: src/Intervals.php (hunk 9 above)
Comment (naderman) [reply src/Intervals.php]: Why did you mark this resolved?

File: src/Intervals.php [hunk 14]
```diff
@@ -0,0 +1,369 @@
...
+
+            // != dev-foo means any numeric version may match
+            if ($op === '!=') {
+                $intervals[] = array('start' => self::zero(), 'end' => self::positiveInfinity());
+            }
+
+            return array('intervals' => $intervals, 'devConstraints' => array($constraint));
+        }
+
+        if ($op[0] === '>') { // > & >=
```
Comment (naderman) [inline src/Intervals.php:315]: as this doesn't handle dev- versions at all, I think this same PR should introduce exceptions if you try to use branch names together with operators other than !=/== ?

Comment (naderman) [reply src/Intervals.php:315]: Nevermind, I missed that those are all covered above already.

File: src/Intervals.php [hunk 15]
```diff
@@ -0,0 +1,369 @@
...
+                            $dev = array();
+                            break 2;
+                        }
+                        // != dev-foo || == dev-foo -> *
+                        if ($op === '==' && $c->getVersion() === $c2->getVersion()) {
+                            $dev = array();
+                            break 2;
+                        }
+                        // != dev-foo || == dev-master -> != dev-foo
+                        if ($op === '==') {
```
Comment (naderman) [inline src/Intervals.php:186]: this behaves incorrectly for anyDev, e.g. `!= dev-foo || == dev*` -> `*` not `!= dev-foo`

Comment (naderman) [reply src/Intervals.php:186]: Should probably also add a test that covers this situation.

File: src/Intervals.php [hunk 16]
```diff
@@ -0,0 +1,369 @@
...
+                            unset($dev[$j]);
+                            continue;
+                        }
+                    }
+                }
+            }
+        } else {
+            $blacklist = array();
```
Comment (naderman) [inline src/Intervals.php:194]: `$disallowlist`

File: src/Intervals.php (hunk 9 above)
Comment (Seldaek) [reply src/Intervals.php]: Oh because I reused the array for another type of sorting at some point but I ended up dropping that. Will remove == and != too I guess can safely go.

File: src/Intervals.php (hunk 13 above)
Comment (Seldaek) [reply src/Intervals.php:277]: With normalized constraints no it is not needed I believe. We do have some tests using non-normalized constraints, and technically nothing preventing you from creating Constraint instances with non-normalized versions. I don't know if we want to just document this as "voiding warranty" and drop the version_compare for perf, or keep supporting these edge cases.

File: src/Intervals.php [hunk 17]
```diff
@@ -0,0 +1,369 @@
...
+                        }
+
+                        foreach ($group2 as $j2 => $c2) {
+                            if ((string) $c2 === (string) $c || $c2 === self::anyDev()) {
+                                $otherGroupMatches++;
+                            }
+
+                            // != x && == x cancel each other, make sure none of these appears in the output
+                            if ($c->getOperator() === '==' && $c2->getOperator() === '!=' && $c->getVersion() === $c2->getVersion()) {
+                                $blacklist[] = (string) $c;
```
Comment (naderman) [inline src/Intervals.php:216]: should probably also be a map rather than array, so you don't unset the same thing multiple times.

File: src/Intervals.php (hunk 13 above)
Comment (Seldaek) [reply src/Intervals.php:277]: BTW the failing tests that triggered me to add this were https://github.com/composer/semver/blob/master/tests/Constraint/ConstraintTest.php#L494 where b2 and beta2 do not compare equally without version_compare. Constraint::matches uses version_compare so technically it is supported to some extent, even though in Composer we never create non-normalized constraints AFAIK.

File: src/Intervals.php (hunk 15 above)
Comment (Seldaek) [reply src/Intervals.php:189]: Fixed in https://github.com/composer/semver/pull/97/commits/3d72ad3c44db65b986e54f0a9ff818beb8c6e066

**Comment by Seldaek** [comment]
Fixed all feedback.

File: src/Intervals.php [hunk 18]
```diff
@@ -127,12 +127,12 @@ public static function get(ConstraintInterface $constraint)
...
 
     /**
-     * @phpstan-return array{'intervals': array<int, array{'start': Constraint, 'end': Constraint}>, 'devConstraints': Constraint[]}
+     * @phpstan-return array{'intervals': Interval[], 'devConstraints': Constraint[]}
      */
     private static function generateIntervals(ConstraintInterface $constraint, $stopOnFirstValidInterval = false)
     {
         if ($constraint instanceof EmptyConstraint) {
-            return array('intervals' => array(array('start' => self::zero(), 'end' => self::positiveInfinity())), 'devConstraints' => array(self::anyDev()));
+            return array('intervals' => array(new Interval(Interval::zero(), Interval::positiveInfinity())), 'devConstraints' => array(Interval::anyDev()));
```
Comment (Toflar) [inline src/Intervals.php]: Imho we have the same optimization possibility here as we'd have with a `ConstraintFactory`: we don't need a new instance of `Interval` over and over again if the interval is exactly the same. We could reuse existing interval instances, right? :)

Comment (Seldaek) [reply src/Intervals.php]: Sure, but IMO that's still best done with the memoization on get(), because the expensive stuff here is computing intervals for a constraint, not storing them. Especially as they'll get used at poolbuilder time and then it can be cleared as it becomes useless. Feel free to PR if you have time though :P
